hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/source/IncrementalInputSplits.java

```diff
@@ -128,12 +128,14 @@ public Result inputSplits(
       return Result.EMPTY;
     }
 
+    // The value may be 'earliest' or Null or outOfRange.
     final String startCommit = this.conf.getString(FlinkOptions.READ_START_COMMIT);
-    final String endCommit = this.conf.getString(FlinkOptions.READ_END_COMMIT);
```

Comment:
Just fix the `startCommit` should work as expected, no need to change to much logic.

Reply:
Now, our problem is that when read-start-commit is outOfRange, no matter what the value of read.end-commit is set, the whole table will be scanned, which causes some redundant data to be scanned and finally leads to incorrect results. Therefore, I think it is necessary to scan the entire table only when the ((start-commit value is null / outofrange / earliest) & & (end-commit value is null / outofrange). (The current logic is:  startFromEarliest || startOutOfRange || endOutOfRange)

Reply:
https://issues.apache.org/jira/browse/HUDI-4726